### PR TITLE
Fixes for usage of use-ssl option

### PR DESCRIPTION
**Pull request by splisson-altair**
- Elasticsearch url generation depending on useSSL value.
- Cerebro configuration.

Fix (partial) for #161

### CODE REVIEW

**Comment by gianrubio** [comment]
@splisson-altair this is still a breaking change, if users bump the operator version and the flag `use-ssl` is not configured in the `CRD` golang will set values as false. You should set the values of  `use-ssl` as true if it's not yet configured in the `CRD`.

**Comment by splisson-altair** [comment]
@gianrubio I agree but I have not figured out how to detect that a value is not set in the CRD vs it being defined to false. From what I found unless one use a constructor, it's not possible to detect that in the struct. Any idea ?

**Comment by splisson-altair** [comment]
Maybe replace use-ssl by http-scheme, string "http" or "https" , that I can test for empty string and default to https

**Comment by SEJeff** [comment]
Given your latest change, this should not be a breaking change to upgrade the operator, correct?

**Comment by splisson-altair** [comment]
@SEJeff It should not. But I was only able to test with 2 configs (with and without) on my k8s cluster, so I'd say it should not break but to be tested more thoroughly.

**Comment by stevesloka** [comment]
This errors out for me when I run:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xe1880b]

goroutine 53 [running]:
github.com/upmc-enterprises/elasticsearch-operator/pkg/k8sutil.(*K8sutil).CreateClientDeployment(0xc4201c2720, 0xc4205320c0, 0x37, 0xc42041e118, 0xc4202b2440, 0x11, 0x0, 0x0, 0x0, 0x0, ...)
	/home/steve/godev/src/github.com/upmc-enterprises/elasticsearch-operator/pkg/k8sutil/deployments.go:121 +0x8eb
github.com/upmc-enterprises/elasticsearch-operator/pkg/processor.(*Processor).processElasticSearchCluster(0xc4201da9c0, 0xc42041e000, 0xc42007fe68, 0x1)
	/home/steve/godev/src/github.com/upmc-enterprises/elasticsearch-operator/pkg/processor/processor.go:298 +0x835
github.com/upmc-enterprises/elasticsearch-operator/pkg/processor.(*Processor).processElasticSearchClusterEvent(0xc4201da9c0, 0xc42041e000, 0x0, 0x0)
	/home/steve/godev/src/github.com/upmc-enterprises/elasticsearch-operator/pkg/processor/processor.go:230 +0x118
github.com/upmc-enterprises/elasticsearch-operator/pkg/processor.(*Processor).WatchElasticSearchClusterEvents.func1(0xc4200460c0, 0xc4201da9c0, 0xc4200a1740, 0xc4200b8420, 0xc42032d150)
	/home/steve/godev/src/github.com/upmc-enterprises/elasticsearch-operator/pkg/processor/processor.go:85 +0x1f8
created by github.com/upmc-enterprises/elasticsearch-operator/pkg/processor.(*Processor).WatchElasticSearchClusterEvents
	/home/steve/godev/src/github.com/upmc-enterprises/elasticsearch-operator/pkg/processor/processor.go:81 +0x78
exit status 2
```

**Comment by splisson-altair** [comment]
@stevesloka I added pointer nil checks that should fix that

**Comment by komljen** [comment]
I did some tests.

Setting `use-ssl: false` worked well. The cluster was running without SSL and health checks passed. I was able to connect from client without any issues.

Setting `use-ssl: true` also worked fine, cluster was deployed with SSL and health checks passed again.

If I omitt `use-ssl`, default was false and it should be set to true?
I see the this message in operator, but still cluster is deployed without SSL:
```
time="2018-04-26T07:30:48Z" level=info msg="use-ssl not specified, defaulting to UseSSL=true"
```

**Comment by stevesloka** [comment]
@komljen thanks for testing, I think defaulting to `true` for ssl would be best. @splisson-altair do you want to make that change? If not we could merge and do a followup PR.

**Comment by splisson-altair** [comment]
Thanks. I ll look into it

Sébastien

On Apr 26, 2018, at 05:58, Steve Sloka <notifications@github.com<mailto:notifications@github.com>> wrote:


@komljen<https://github.com/komljen> thanks for testing, I think defaulting to true for ssl would be best. @splisson-altair<https://github.com/splisson-altair> do you want to make that change? If not we could merge and do a followup PR.

**Comment by splisson-altair** [comment]
An assignment was missing for a new cluster. The cluster spec receive din the function is used for new cluster, not the one updated in refreshCluster

File: pkg/processor/processor_test.go [hunk 1]
```diff
@@ -111,3 +113,24 @@ func TestNoZonesPassed(t *testing.T) {
 		t.Error("Expected 1 zone, got ", len(zoneDistribution))
 	}
 }
+
+func TestDefaultUseSSL(t *testing.T) {
+	processor, _ := New(nil, "foo/image")
+
+	useSSL := processor.defaultUseSSL(nil)
+	if useSSL != true {
+		t.Error("Expected useSSL to default to true when not specified, got %v", useSSL)
```
Comment (stevesloka) [inline pkg/processor/processor_test.go]: Should be `Errorf`

File: pkg/processor/processor_test.go [hunk 2]
```diff
@@ -111,3 +113,24 @@ func TestNoZonesPassed(t *testing.T) {
...
+	}
+
+	useSSL = true
+	useSSL = processor.defaultUseSSL(&useSSL)
+	if useSSL != true {
+		t.Error("Expected useSSL to be true when specified as true, got %v", useSSL)
```
Comment (stevesloka) [inline pkg/processor/processor_test.go]: Should be `Errorf`

File: pkg/processor/processor_test.go [hunk 3]
```diff
@@ -111,3 +113,24 @@ func TestNoZonesPassed(t *testing.T) {
...
+	}
+
+	useSSL = false
+	useSSL = processor.defaultUseSSL(&useSSL)
+	if useSSL != false {
+		t.Error("Expected useSSL to be false when specified as false, got %v", useSSL)
```
Comment (stevesloka) [inline pkg/processor/processor_test.go]: Should be `Errorf`

**Comment by stevesloka** [comment]
I set `use-ssl=false` and it still created certs:

```
time="2018-04-26T20:28:19Z" level=info msg="use-ssl false"
time="2018-04-26T20:28:19Z" level=info msg="Using [upmcenterprises/docker-elasticsearch-kubernetes:6.1.3_0] as image for es cluster"
time="2018-04-26T20:28:19Z" level=info msg="use-ssl false"
time="2018-04-26T20:28:19Z" level=info msg="Creating new certs!"
time="2018-04-26T20:28:19Z" level=info msg="Creating ca cert..."
time="2018-04-26T20:28:19Z" level=info msg="Creating node cert..."
time="2018-04-26T20:28:19Z" level=info msg="Creating kibana cert..."
time="2018-04-26T20:28:20Z" level=info msg="Creating cerebro cert..."
time="2018-04-26T20:28:21Z" level=info msg="Converting node to pkcs12..."
time="2018-04-26T20:28:21Z" level=info msg="Converting ca cert to jks..."
time="2018-04-26T20:28:21Z" level=info msg="Converting node cert to jks..."
time="2018-04-26T20:28:21Z" level=info msg="Discovery Service not found, creating..."
```

File: pkg/processor/processor_test.go (hunk 1 above)
Comment (splisson-altair) [reply pkg/processor/processor_test.go]: Thanks

**Comment by splisson-altair** [comment]
Yes it will require to modify the way the StatefulSet struct is built.
I'm also looking into that area of the code to allow updating Resources requests and limits. AFAIK currently one need to delete the cluster to do that with the operator.

**Comment by stevesloka** [comment]
I'm going to merge this, we can fix the certs piece in another PR

**Pull request merged**